### PR TITLE
[trainer] fix issues where number of failures would always be zero

### PR DESCRIPTION
**Pull request by tahirmt**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I see several green `ci/circleci` builds in the "All checks have passed" section of my PR ([connect CircleCI to GitHub](https://support.circleci.com/hc/en-us/articles/360008097173-Why-aren-t-pull-requests-triggering-jobs-on-my-organization-) if not)
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

Closes #21565

### Description
If the test case name has spaces or any special characters, the reported `number_of_failures` will always be 0. This will also not include any failed tests inside the junit report.

The reason this was happening was because during the comparison to check the test failure there was a sanitization step that's not applied to the identifier itself so it never matched.

This is from the log output of such a test.
```
** TEST FAILED **
[15:11:03]: Exit status: 65
+--------------------+---+
|      Test Results      |
+--------------------+---+
| Number of tests    | 3 |
| Number of failures | 0 |
+--------------------+---+

[15:11:04]: Test execution failed. Exit status: 65
```

### Testing Steps

The easiest way to reproduce it was to use https://github.com/Quick/quick because Quick 7 produces test case names with spaces if the spec names had spaces in it. And when you run `scan` the output will always say the number of failures is 0.

I have attached a sample project.

1. Download project
2. Run `bundle install`
3. Run `bundle exec fastlane test` 

When it runs the output will show number of failures as 0.

Then you can uncomment the branch line in the `Gemfile` and run bundle install and switch to this branch. Then you will see that the number of failure is 1.

[SpaceTests.zip](https://github.com/fastlane/fastlane/files/12221320/SpaceTests.zip)

### CODE REVIEW

**Comment by tahirmt** [comment]
I'm not sure how to create a sample xcresult file with test name with spaces. I tried a sample project but that created a 50MB file which feels too large to add to the project

**Comment by tahirmt** [comment]
The failure is unrelated to my changes. Likely existing issue.

**Comment by tahirmt** [comment]
@joshdholtz any chance you can take a look at this?

**Review by rogerluan — APPROVED** [review]
Thank you so much for providing a unit test, a fixture .xcresult and a sample project @tahirmt !

I reviewed your solution and it was great, but I thought of merging in the approach that @TheMetalCode applied in his PR that addresses this very same issue https://github.com/fastlane/fastlane/pull/21565 (but his PR was lacking unit tests 😅). In summary, I replaced the "String extension" that you created with a simple `proc`, and changed the implementation of the sanitizer from using several individual custom string replacements, with a single regex-replace that looks for all `\W` (i.e. non-word characters).

I think this is more fail-proof, and passes all the tests 😊 

This PR LGTM already. I've made the changes above and solved some merge conflicts that were present. I'll give a few more days before merging it to see if there's anyone else interested in reviewing it too 🙇 maybe @lacostej ?

Thanks once again for this @tahirmt , this is a long-lasting issue! 🙏

**Comment by rogerluan** [comment]
PS: @tahirmt let me know what are your thoughts on the changes I made 🙇

**Comment by tahirmt** [comment]
@rogerluan the solution looks good to me. Thanks for updating it and moving it forward.

**Comment by rogerluan** [comment]
No problem! Looking forward to merging this soon!

**Review by lacostej — APPROVED** [review]
LGTM

**Comment by tahirmt** [comment]
@rogerluan any idea when we can expect it to be merged?

**Comment by rogerluan** [comment]
Hi @tahirmt, I'm OOO till Sunday, but I'll take a look once I'm back :)

**Comment by rogerluan** [comment]
Thanks for another great contribution to _fastlane_ @tahirmt 🤗 this is going in now! Thank you for your patience 🙇

**Comment by tahirmt** [comment]
@rogerluan sorry to bug you but do you have any idea when the next fastlane release would go out? It's been weeks since it was merged but no release yet.

**Comment by rogerluan** [comment]
Nope, sorry 😬 I don't have control over the releases yet 😅 I'll try to get a release out soon though 🙏

**Comment by jaysoffian** [comment]
This change actually broke parsing for at least some result files:

Screenshot 2024-04-19 at 5 33 57 PM

If I can't figure this out myself, I'll attach the sanitized output of `xcrun xcresulttool get --format json` for the xcresult bundle in question. I'm just noting this PR introduced a regression.

Result bundle was generated by Xcode 15.1.

File: trainer/lib/trainer/xcresult.rb [hunk 1]
```diff
@@ -184,16 +186,10 @@ def find_failure(failures)
...
-                                              .tr(".", "/")
-                                              .tr("-", "")
-                                              .tr("[", "")
-                                              .tr("]", "")
-                                              .tr(" ", "/")
-            self.identifier == sanitized_test_case_name
+            # Sanitize both test case name and identifier in a consistent fashion, then replace all non-word
+            # chars with underscore, and compare them
+            sanitized_test_case_name = sanitizer.call(failure.test_case_name)
+            sanitized_identifier == sanitized_test_case_name
```
Comment (jaysoffian) [inline trainer/lib/trainer/xcresult.rb:192]: This change is a regression. See https://github.com/fastlane/fastlane/pull/21432#issuecomment-2067310932

Comment (jaysoffian) [reply trainer/lib/trainer/xcresult.rb:192]: Fix https://github.com/fastlane/fastlane/pull/21990

**Pull request merged**